include/ginkgo/core/base/math.hpp

```diff
@@ -200,6 +305,94 @@ GKO_INLINE GKO_ATTRIBUTES constexpr bool is_complex()
 }
 
 
+/**
+ * Allows to check if T is a complex or scalar value during compile time by
+ * accessing the `value` attribute of this struct. If `value` is `true`, T is a
+ * complex/scalar type, if it is `false`, T is not a complex/scalar type.
+ *
+ * @tparam T  type to check
+ */
+template <typename T>
+using is_complex_or_scalar_s = detail::is_complex_or_scalar_impl<T>;
+
+/**
+ * Checks if T is a complex/scalar type.
+ *
+ * @tparam T  type to check
+ *
+ * @return `true` if T is a complex/scalar type, `false` otherwise
+ */
+template <typename T>
+GKO_INLINE GKO_ATTRIBUTES constexpr bool is_complex_or_scalar()
+{
+    return detail::is_complex_or_scalar_impl<T>::value;
+}
+
+
+/**
+ * Removes the complex of complex/scalar type or the template parameter of class
+ * by accessing the `type` attribute of this struct.
+ *
+ * @tparam T  type to remove complex
+ */
+using detail::remove_complex_s;
+
+/**
+ * Obtain the type which removed the complex of complex/scalar type or the
+ * template parameter of class by accessing the `type` attribute of this struct.
+ *
+ * @tparam T  type to remove complex
+ *
+ * @note remove_complex<class> can not be used in friend class declaration.
+ */
+template <typename T>
+using remove_complex = typename remove_complex_s<T>::type;
+
+
+/**
+ * Add the complex of complex/scalar type or the template parameter of class
+ * by accessing the `type` attribute of this struct.
+ *
+ * @tparam T  type to complex_type
+ */
+using detail::to_complex_s;
+
+/**
+ * Obtain the type which adds the complex of complex/scalar type or the
+ * template parameter of class by accessing the `type` attribute of this struct.
+ *
+ * @tparam T  type to complex_type
+ *
+ * @note to_complex<class> can not be used in friend class declaration.
+ *       the followings are the error message from different combination.
+ *       friend to_complex<Csr>;
+ *         error: can not recognize it is class correctly.
+ *       friend class to_complex<Csr>;
+ *         error: using alias template specialization
+ *       friend class to_complex_s<Csr<ValueType,IndexType>>::type;
+ *         error: can not recognize it is class correctly.
+ */
+template <typename T>
+using to_complex = typename to_complex_s<T>::type;
+
+
+/**
+ * to_real_s is alias of remove_complex_s
+ *
+ * @tparam T  type to real
+ */
+template <typename T>
+using to_real_s = remove_complex_s<T>;
```

Comment:
Where would we need this (I mean the `to_real_s` variant in `gko` namespace)? `to_real` should be enough IMO.